### PR TITLE
Legacy: simple classif training

### LINKED ISSUE #5491: Improve legacy checkpoint testing
## 🚀 Feature

Improve legacy checkpoint testing

### Motivation

Currently, we only test whether a legacy checkpoint can be loaded and run, but no internal state is checked.

### Pitch

Check the epoch number is correct, if the state is the expected, if we achieve a particular loss after 1 step...

We already have similar tests, we just have to do them for legacy checkpoints. see

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/models/test_restore.py

### Additional context

https://github.com/PyTorchLightning/pytorch-lightning/blob/1f6236accce78303249c55de656b71501e607d1a/tests/checkpointing/test_legacy_checkpoints.py#L49-L72

cc: @Borda

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Resolves #5491

## Before submitting
- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [ ] Did you list all the **breaking changes** introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/checkpointing/test_legacy_checkpoints.py [hunk 1]
```diff
@@ -14,84 +14,118 @@
...
 
-from pytorch_lightning import Trainer
+import pytorch_lightning as pl
+from pytorch_lightning import Callback, Trainer
+from pytorch_lightning.callbacks import EarlyStopping
 from tests import _PATH_LEGACY
 
 LEGACY_CHECKPOINTS_PATH = os.path.join(_PATH_LEGACY, 'checkpoints')
 CHECKPOINT_EXTENSION = ".ckpt"
+LEGACY_BACK_COMPATIBLE_PL_VERSIONS = (
```
Comment (Borda) [inline tests/checkpointing/test_legacy_checkpoints.py:29]: I have not populated all past legacy, so all legacy test are failing as they are expecting the actual model
once I do it and upload to S3, merging this PR become p0 - blocker

**Comment by Borda** [comment]
new generated for local testing... [checkpoints.zip](https://github.com/PyTorchLightning/pytorch-lightning/files/6868667/checkpoints.zip)

File: tests/checkpointing/test_legacy_checkpoints.py [hunk 2]
```diff
@@ -14,84 +14,117 @@
...
+            max_epochs=21,
+            accumulate_grad_batches=2,
+            deterministic=True,
+            resume_from_checkpoint=path_ckpt
+        )
+        trainer.fit(model, datamodule=dm)
+        res = trainer.test(model, datamodule=dm)
+        assert res[0]['test_loss'] <= 0.7
+        assert res[0]['test_acc'] >= 0.85
+        print(res)
```
Comment (carmocca) [inline tests/checkpointing/test_legacy_checkpoints.py:130]: What is the point of printing here?

If the assertions fail, this will not appear anyways

File: tests/checkpointing/test_legacy_checkpoints.py [hunk 3]
```diff
@@ -14,84 +14,117 @@
...
-    trainer = Trainer(default_root_dir=tmpdir, max_epochs=6)
-    trainer.fit(model)
+    def __init__(self, nb: int):
+        self.limit = nb
+        self._count = 0
+
+    def on_epoch_start(self, trainer: 'pl.Trainer', pl_module: 'pl.LightningModule') -> None:
+        self._count += 1
+        if self._count >= self.limit:
+            trainer.should_stop = True
```
Comment (carmocca) [inline tests/checkpointing/test_legacy_checkpoints.py:98]: Are you sure this still works on the oldest versions?

File: tests/checkpointing/test_legacy_checkpoints.py [hunk 4]
```diff
@@ -14,84 +14,117 @@
...
+    PATH_LEGACY = os.path.join(LEGACY_CHECKPOINTS_PATH, pl_version)
+    with patch('sys.path', [PATH_LEGACY] + sys.path):
+        from simple_classif_training import ClassifDataModule, ClassificationModel
 
-    # todo
-    # model = DummyModel()
-    # trainer = Trainer(default_root_dir=tmpdir, max_epochs=1, resume_from_checkpoint=path_ckpt)
-    # trainer.fit(model)
+        path_ckpts = sorted(glob.glob(os.path.join(PATH_LEGACY, f'*{CHECKPOINT_EXTENSION}')))
+        assert path_ckpts, 'No checkpoints found in folder "%s"' % PATH_LEGACY
```
Comment (carmocca) [inline tests/checkpointing/test_legacy_checkpoints.py:108]:
```suggestion
        assert path_ckpts, f'No checkpoints found in folder "{PATH_LEGACY}"'
```

File: tests/checkpointing/test_legacy_checkpoints.py [hunk 5]
```diff
@@ -14,84 +14,117 @@
...
-        "1.3.5",
-        "1.3.6",
-        "1.3.7",
-        "1.3.8",
-    ]
-)
-def test_resume_legacy_checkpoints(tmpdir, pl_version: str):
-    path_dir = os.path.join(LEGACY_CHECKPOINTS_PATH, pl_version)
+@pytest.mark.parametrize("pl_version", LEGACY_BACK_COMPATIBLE_PL_VERSIONS)
+def test_load_legacy_checkpoints(tmpdir, pl_version: str):
```
Comment (carmocca) [inline tests/checkpointing/test_legacy_checkpoints.py:71]: What is the difference supposed to be between the tests

File: legacy/simple_classif_training.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
+# Copyright The PyTorch Lightning team.
```
Comment (carmocca) [inline legacy/simple_classif_training.py:1]: What is the point of this file + the other legacy checkpoints?

File: tests/checkpointing/test_legacy_checkpoints.py (hunk 2 above)
Comment (Borda) [reply tests/checkpointing/test_legacy_checkpoints.py:130]: Just to see the margin, will move it up

File: tests/checkpointing/test_legacy_checkpoints.py (hunk 3 above)
Comment (Borda) [reply tests/checkpointing/test_legacy_checkpoints.py:98]: Seems so, tested it on master

File: tests/checkpointing/test_legacy_checkpoints.py (hunk 5 above)
Comment (Borda) [reply tests/checkpointing/test_legacy_checkpoints.py:71]: One is load from checkpoint other is resume, or what are you asking?

File: legacy/simple_classif_training.py (hunk 6 above)
Comment (Borda) [reply legacy/simple_classif_training.py:1]: What do you mean?
This is template for creating checkpoint, this one is for master, the copies in version folders are in case some model chnages or needed adjustments to be runnable

File: tests/checkpointing/test_legacy_checkpoints.py (hunk 5 above)
Comment (carmocca) [reply tests/checkpointing/test_legacy_checkpoints.py:71]: But do you expect there to be a difference between them?
`resume_from_checkpoint` loads from a checkpoint

Comment (Borda) [reply tests/checkpointing/test_legacy_checkpoints.py:71]: shall not, but still rather test it, then just hope :]

File: tests/checkpointing/test_legacy_checkpoints.py [hunk 7]
```diff
@@ -14,84 +14,117 @@
...
+    PATH_LEGACY = os.path.join(LEGACY_CHECKPOINTS_PATH, pl_version)
+    with patch('sys.path', [PATH_LEGACY] + sys.path):
+        from simple_classif_training import ClassifDataModule, ClassificationModel
+
+        path_ckpts = sorted(glob.glob(os.path.join(PATH_LEGACY, f'*{CHECKPOINT_EXTENSION}')))
+        assert path_ckpts, 'No checkpoints found in folder "%s"' % PATH_LEGACY
```
Comment (Borda) [inline tests/checkpointing/test_legacy_checkpoints.py:77]:
```suggestion
        assert path_ckpts, f'No checkpoints found in folder "{PATH_LEGACY}"'
```

File: legacy/simple_classif_training.py (hunk 6 above)
Comment (Borda) [reply legacy/simple_classif_training.py:1]: or would you suggest o use the same source code for all PL versions?

File: legacy/README.md [hunk 8]
```diff
@@ -12,6 +12,6 @@ unzip -o checkpoints.zip
 
 To back populate collection with past version you can use following bash:
 ```bash
-bash generate_checkpoints.sh 1.0.2 1.0.3 1.0.4
+bash generate_checkpoints.sh "1.0.0" "1.0.1" "1.0.2" "1.0.3" "1.0.4" "1.0.5" "1.0.6" "1.0.7" "1.0.8" "1.1.0" "1.1.1" "1.1.2" "1.1.3" "1.1.4" "1.1.5" "1.1.6" "1.1.7" "1.1.8" "1.2.0" "1.2.1" "1.2.2" "1.2.3" "1.2.4" "1.2.5" "1.2.6" "1.2.7" "1.2.8" "1.2.10" "1.3.0" "1.3.1" "1.3.2" "1.3.3" "1.3.4" "1.3.5" "1.3.6" "1.3.7" "1.3.8"
```
Comment (tchaton) [inline legacy/README.md:15]: We can auto-generate this list with `pip` directly ?

File: legacy/simple_classif_training.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+    def configure_optimizers(self):
+        optimizer = torch.optim.Adam(self.parameters(), lr=self.lr)
+        return [optimizer], []
+
+    def training_step(self, batch, batch_idx):
+        x, y = batch
+        logits = self.forward(x)
+        loss = F.cross_entropy(logits, y)
+        self.log('train_loss', loss, prog_bar=True)
```
Comment (tchaton) [inline legacy/simple_classif_training.py:131]: Hey @Borda, quick questions to improve this test for stability.

Should we check the loss values to ensure consistency ?
Should we test manual optimization ?
Should we test Callback states restoration ?

File: legacy/README.md (hunk 8 above)
Comment (Borda) [reply legacy/README.md:15]: this is just temp, for me generating all past, the automated takes only actual release

File: legacy/simple_classif_training.py (hunk 9 above)
Comment (Borda) [reply legacy/simple_classif_training.py:131]: > Should we check the loss values to ensure consistency?

now we check `test_loss` so you suggest testing all three losses - train/val/test?

> Should we test manual optimization?

do you mean to add another model with manual optimization?

> Should we test Callback states restoration?

That would be good but it requires to have different source code per past version as not all callbacks were present since 1.0 yet... so lets make it a separate PR

File: legacy/generate_checkpoints.sh [hunk 10]
```diff
@@ -2,12 +2,14 @@
 # Sample call:
 #  bash generate_checkpoints.sh 1.0.2 1.0.3 1.0.4
 
+set -e
+
 LEGACY_PATH="$( cd "$(dirname "$0")" >/dev/null 2>&1 ; pwd -P )"
 FROZEN_MIN_PT_VERSION="1.4"
```
Comment (carmocca) [inline legacy/generate_checkpoints.sh:8]: Should this be updated?

File: legacy/simple_classif_training.py (hunk 9 above)
Comment (tchaton) [reply legacy/simple_classif_training.py]: 1. Should we check the loss values to ensure consistency?

Here is what I meant.

```
def test_legacy()

    target_losses = [1.2, 1....] # recorded once and used to ensure all models converges in the same way after each release

    for chk in checkpoints:
        seed_everything(42)
        losses = evaluate(chk)
        assert losses == previous_losses
```

2. do you mean to add another model with manual optimization? Yes.

3. Yes, we can start from the lowest version properly supporting the callback states. Definitely within another PR.

Comment (Borda) [reply legacy/simple_classif_training.py]: > 1. Should we check the loss values to ensure consistency?
> 
> Here is what I meant.
> 
> ```
> def test_legacy()
> 
>     target_losses = [1.2, 1....] # recorded once and used to ensure all models converges in the same way after each release
> 
>     for chk in checkpoints:
>         seed_everything(42)
>         losses = evaluate(chk)
>         assert losses == previous_losses
> ```

on, that would need to save the values so lets make it another PR
 
> 2. do you mean to add another model with manual optimization? Yes.

Would need to change the logic for finding ckpt and name then, so another PR

> 3. Yes, we can start from the lowest version properly supporting the callback states. Definitely within another PR.

another PR :rabbit:

File: tests/checkpointing/test_legacy_checkpoints.py [hunk 11]
```diff
@@ -14,84 +14,116 @@
...
 
-    sys.path = orig_sys_paths
+        dm = ClassifDataModule()
+        model = ClassificationModel()
+        es = EarlyStopping(monitor="val_acc", mode="max", min_delta=0.005)
+        stop = LimitNbEpochs(1)
+        trainer = Trainer(
+            default_root_dir=str(tmpdir),
+            gpus=int(torch.cuda.is_available()),
+            precision=16 if torch.cuda.is_available() else 32,
```
Comment (awaelchli) [inline tests/checkpointing/test_legacy_checkpoints.py:117]:
```suggestion
            precision=(16 if torch.cuda.is_available() else 32),
```

**Review by awaelchli — APPROVED** [review]

File: legacy/simple_classif_training.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+        self.log("test_acc", self.test_acc(logits, y), prog_bar=True)
+
+
+def main_train(dir_path, max_epochs: int = 20):
+    seed_everything(42)
+    stopping = EarlyStopping(monitor="val_acc", mode="max", min_delta=0.005)
+    trainer = pl.Trainer(
+        default_root_dir=dir_path,
+        gpus=int(torch.cuda.is_available()),
+        precision=16 if torch.cuda.is_available() else 32,
```
Comment (Borda) [inline legacy/simple_classif_training.py:150]:
```suggestion
        precision=(16 if torch.cuda.is_available() else 32),
```

File: legacy/simple_classif_training.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        max_epochs=max_epochs,
+        accumulate_grad_batches=2,
+        deterministic=True,
+    )
+
+    dm = ClassifDataModule()
+    model = ClassificationModel()
+    trainer.fit(model, datamodule=dm)
+    res = trainer.test(model, datamodule=dm)
+    print(res)
```
Comment (tchaton) [inline legacy/simple_classif_training.py:163]: Is it required ?

File: tests/checkpointing/test_legacy_checkpoints.py [hunk 14]
```diff
@@ -14,84 +14,116 @@
...
+            max_epochs=21,
+            accumulate_grad_batches=2,
+            deterministic=True,
+            resume_from_checkpoint=path_ckpt,
+        )
+        trainer.fit(model, datamodule=dm)
+        res = trainer.test(model, datamodule=dm)
+        assert res[0]["test_loss"] <= 0.7
+        assert res[0]["test_acc"] >= 0.85
+        print(res)
```
Comment (tchaton) [inline tests/checkpointing/test_legacy_checkpoints.py:129]: Same here ?

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by tchaton — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]
